xscen/diagnostics.py

```diff
@@ -2,110 +2,290 @@
+            ds_start = ds.time.min().values
+            ds_end = ds.time.max().values
+    if start_date is not None:
+        # Create cf_time objects to compare the dates
+        start_date = xr.cftime_range(
+            start=start_date, periods=1, freq="D", calendar=ds.time.dt.calendar
+        )[0]
+        if not ((ds_start <= start_date) and (ds_end > start_date)):
+            err = f"The start date is not at least {start_date}. Received {ds.time.min().values.astype('datetime64[m]')}."
+            if "start_date" in raise_on:
+                raise ValueError(err)
+            else:
+                warnings.warn(err, UserWarning, stacklevel=1)
+    if end_date is not None:
+        # Create cf_time objects to compare the dates
+        end_date = xr.cftime_range(
+            start=end_date, periods=1, freq="D", calendar=ds.time.dt.calendar
+        )[0]
+
+        if not ((ds_start < end_date) and (ds_end >= end_date)):
+            err = f"The end date is not at least {end_date}. Received {ds.time.max().values.astype('datetime64[m]')}."
+            if "end_date" in raise_on:
+                raise ValueError(err)
+            else:
+                warnings.warn(err, UserWarning, stacklevel=1)
+
+    # Check variables
+    if variables_and_units is not None:
+        for v in variables_and_units:
+            if v not in ds:
+                raise ValueError(f"The variable '{v}' is missing.")
+            if ds[v].attrs.get("units", None) != variables_and_units[v]:
+                xc.core.units.check_units(
+                    ds[v], variables_and_units[v]
+                )  # Will always raise an error if the units are not compatible
+                err = f"The variable '{v}' does not have the expected units '{variables_and_units[v]}'. Received '{ds[v].attrs['units']}'."
+                if "variables_and_units" in raise_on:
+                    raise ValueError(err)
+                else:
+                    warnings.warn(err, UserWarning, stacklevel=1)
+
+    # Check CF conventions
+    if cfchecks is not None:
+        cfchecks = deepcopy(cfchecks)
+        for v in cfchecks:
+            for check in cfchecks[v]:
+                if check == "check_valid":
+                    cfchecks[v][check]["var"] = ds[v]
+                elif check == "cfcheck_from_name":
+                    cfchecks[v][check].setdefault("varname", v)
+                    cfchecks[v][check]["vardata"] = ds[v]
+                else:
+                    raise ValueError(f"Check '{check}' is not in xclim.")
+                if "cfchecks" in raise_on:
+                    warnings.simplefilter("error")
+                    try:
+                        getattr(xc.core.cfchecks, check)(**cfchecks[v][check])
+                    except UserWarning as e:
+                        raise ValueError(e)
+                    warnings.resetwarnings()
+                else:
+                    getattr(xc.core.cfchecks, check)(**cfchecks[v][check])
```

Comment:
We shouldn't assume that `warn` is the current xclim option.

Option 1 : more explicit:
```suggestion
                err = f"CF-check error for variable {v}."
                with xc.set_options(cf_compliance='raise'):
                    try:
                        getattr(xc.core.cfchecks, check)(**cfchecks[v][check])
                    except xc.core.utils.ValidationError as e:
                        if "cfchecks" in raise_on:
                            raise ValueError() from e
                        else:
                            warnings.warn(err + ' ' + e.args[0], UserWarning, stacklevel=1)
```
or option 2 : directly pass xclim's error/warning
```python3
                with xc.set_options(cf_compliance='raise' if 'cfchecks' in raise_on else 'warn'):
                    getattr(xc.core.cfchecks, check)(**cfchecks[v][check])
```

Reply:
To ensure that we catch everything, I went with option 1 whenever we call xclim. The only exception is `data_flags`, since that would prevent us from returning `dsflags` as an output of the function.